### PR TITLE
resource: assign ranks in R provided by job-info based on hostlist attribute

### PR DESCRIPTION
**Pull request by grondo**
As discussed in #4182, the job shell may not assign ranks in an order that follows the parent instance rank ordering. This occurs when there is a heterogeneous allocation of cores (i.e. different numbers of cores per execution target). This behavior is a hold over from a different time, and there is probably an argument that the shell should not do this, but there is a slightly better argument that the `hostlist` attribute is the canonical mapping of hosts to ranks for an instance, and R should follow this mapping not its own. (also, providing this solution is much easier than changing the shell behavior for now).

This PR adds support for re-ranking R in the resource module when R is provided by the parent, as it already does when R is provided by a bootstrap configuration. The re-ranking is only attempted when the `hostlist` attribute exists (which it always should), and there are no duplicates in `hostlist` (i.e. we're not running multiple brokers per node). Also, if re-ranking fails, this is not an error, since this can occur in testing when using a fake R. However, the re-ranking should fix the issue described in #4182, since there will always be a valid `hostlist` when the shell reorders hosts during a job launch of a heterogeneous allocation.

For testing purposes, the ability to set a `hostlist` on the command line is also introduced in this PR, and the test simply runs a job with a re-ordered set of hosts and ensures that the `resource.R` attribute in the resulting instance was properly rank-ordered.

### CODE REVIEW

File: src/broker/boot_pmi.c
Comment (garlick) [inline src/broker/boot_pmi.c:174]: `free()` potentially clobbers errno, which needs to be set so `log_err()` prints something sensible on failure.

File: src/modules/resource/inventory.c
Comment (garlick) [inline src/modules/resource/inventory.c:395]: The condutional would be slightly clearer IMHO as
```c
if (hostlist_count (hl) == count)
    result = true;
```

File: src/modules/resource/inventory.c
Comment (garlick) [inline src/modules/resource/inventory.c:435]: s/on/one/

**Review by garlick — APPROVED** [review]
LGTM, just a couple minor things.  Nice job coming up with this fix so quickly.

**Comment by dongahn** [comment]
@grondo: thanks. I don't see any red flag. So https://github.com/flux-framework/flux-sched/pull/916 should wait until this gets merged?

**Comment by grondo** [comment]
>  So flux-framework/flux-sched#916 should wait until this gets merged?

No need to wait IMO.

**Comment by grondo** [comment]
I've addressed @garlick's comments and force-pushed the result.

**Comment by garlick** [comment]
Thanks!  Feel free to set MWP.

**Comment by grondo** [comment]
Thanks!

**Pull request merged**
